virt-launcher, Fix flaky VMI report (random missing Guest Agent data)

oshoval opened the pull request:
`virt-handler` refreshes the VMI domain information,
each 300 seconds by `handleResync` function.
    
The information it got doesn't include
Guest Agent information such as Guest OS Info,
and interfaces status.
   
This will cause flakiness in the VMI status,
each time that the report is fetched,
until the next update of Guest Agent will be sent
by the `virt-launcher` to the `virt-handler`,
as this report does include the Guest Agent data.
   
This PR fix it by decorating the resync
domain info, with the Guest Agent information,
as the `virt-launcher` itself does before sending
the domain info to the `virt-handler`.

Replaces reverted https://github.com/kubevirt/kubevirt/pull/5428
see https://github.com/kubevirt/kubevirt/pull/5557 for more info.

Fixes:
*  https://bugzilla.redhat.com/show_bug.cgi?id=1907707
*  https://bugzilla.redhat.com/show_bug.cgi?id=1924479

```release-note
virt-launcher now populates domain's guestOS info and interfaces status according guest agent also when doing periodic resyncs. 
```
Signed-off-by: Or Shoval <oshoval@redhat.com>

oshoval commented:
/cc @AlonaKaplan

oshoval commented:
Do we want to close the revert, and just focus on this one ?
tested locally with and without the fix, it looks good

maiqueb commented:
> Do we want to close the revert #5557, and just focus on this one ?
> otherwise we will need to represent the reverted code in this one, and have both merged.
> tested locally with and without the fix, it looks good

I'd rather keep the revert, and take the time to improve this PR with a unit test.

For what it's worth, I think this PR addresses the issue elegantly (afaiu, we've overwriting the current interfaces we read from the pod interface cache with an empty array), but would be more comfortable with proof; a simple unit test for the broken scenario (bridge interface, no guest agent) to [vm_test.go](https://github.com/kubevirt/kubevirt/blob/972948ef3591a83deb649c8caad71f648086b426/pkg/virt-handler/vm_test.go) would do it. 

There's a unit test in place that you can use as reference in https://github.com/kubevirt/kubevirt/blob/972948ef3591a83deb649c8caad71f648086b426/pkg/virt-handler/vm_test.go#L2295

I think you should add first the test, assure that it fails, add your fix, assure that it passes.

oshoval commented:
We don't override in case there is no guest agent update, this is the fix.
About the unit tests, I think it should be in the test that check the resync logic,
the resync timer must be long enough in order for it to happen btw.
Example is here
https://github.com/kubevirt/kubevirt/pull/5428/files#diff-6406f1444186b45ac44ceb8e369fd9b6ce0c9a4cc6f313bf80e815f0f33e705bR302

Saying that because from what i understood in the past, unit test scope should be the unit they test, not an end 2 end (including the `updateVMIStatus` which was not changed)

@AlonaKaplan wdyt ?
(asking because we had this conversation in the past)

oshoval commented:
Rebased because the revert was merged,
will squash at the end

maiqueb commented:
> We don't override at all in case there is no guest agent update (at all, which happen only if no guest agent exists or ever reported), this is the fix.
> It does the same on the virt-launcher push notification, and this PR does the same behavior for the periodic updates.
> 
> About the unit tests, I think it should be in the test that check the resync logic,
> the resync timer must be long enough in order for it to happen btw.

In my book, the fact that you need to concern yourself w/ a timer, a timeout and all those shebangs are a disadvantage. 

> Example is here
> https://github.com/kubevirt/kubevirt/pull/5428/files#diff-6406f1444186b45ac44ceb8e369fd9b6ce0c9a4cc6f313bf80e815f0f33e705bR302
> 
> Saying that because from what i understood in the past, unit test scope should be the unit they test, not an end 2 end (including the `updateVMIStatus` which was not changed)

I'm simply pointing out that there isn't a unit test that checks if the the interface status is correct in this scenario. 

As such, that's the test I would like to see: be sure that the interface status is read correctly when the guest agent is nowhere to be found for bridge binding. 

If there's no guest agent what is there to sync ? Could be me, but I do not understand that. 

On the other hand, anyone can expect `vmi.Status.Interfaces` to `BeEqual(expectedInterface) `. That is simple to reason about.

You can of course test the whole tree of missing details in the whole code stack - whatever resyncs included.

> 
> @AlonaKaplan wdyt ?
> (asking because we had this conversation in the past)

oshoval commented:
> > We don't override at all in case there is no guest agent update (at all, which happen only if no guest agent exists or ever reported), this is the fix.
> > It does the same on the virt-launcher push notification, and this PR does the same behavior for the periodic updates.
> > About the unit tests, I think it should be in the test that check the resync logic,
> > the resync timer must be long enough in order for it to happen btw.
> 
> In my book, the fact that you need to concern yourself w/ a timer, a timeout and all those shebangs are a disadvantage.
> 

A timer is needed because the resync is triggered periodically, by the virt-handler itself,
this PR did not introduced this timer, nor the unit test that check it, but it just extended its functionality,

> > Example is here
> > https://github.com/kubevirt/kubevirt/pull/5428/files#diff-6406f1444186b45ac44ceb8e369fd9b6ce0c9a4cc6f313bf80e815f0f33e705bR302
> > Saying that because from what i understood in the past, unit test scope should be the unit they test, not an end 2 end (including the `updateVMIStatus` which was not changed)
> 
> I'm simply pointing out that there isn't a unit test that checks if the the interface status is correct in this scenario.
> 
> As such, that's the test I would like to see: be sure that the interface status is read correctly when the guest agent is nowhere to be found for bridge binding.
> 
> If there's no guest agent what is there to sync ? Could be me, but I do not understand that.
> 
> On the other hand, anyone can expect `vmi.Status.Interfaces` to `BeEqual(expectedInterface) `. That is simple to reason about.
> 
> You can of course test the whole tree of missing details in the whole code stack - whatever resyncs included.

I discussed with Alona offline briefly, and in the past about it, I will start with updating the resync one, since the unit test should test that, 
once its ready we can discuss again.
There are e2e tests that do check the full flow.
The unit test checks the resync logic, if there resync is working as expected, the rest is for the e2e to check it according my view.

> 
> > @AlonaKaplan wdyt ?
> > (asking because we had this conversation in the past)

maiqueb commented:
> I discussed with Alona offline briefly, and in the past about it, I will start with updating the resync one, since the unit test should test that,
> once its ready we can discuss again.
> There are e2e tests that do check the full flow.
> The unit test checks the resync logic, if there resync is working as expected, the rest is for the e2e to check it according my view.

I still think its a mistake not to provide a simple test that can be easily used to reason about this but I can't force anyone to agree with me. 

I can add it in an unrelated PR.

> 
> > > @AlonaKaplan wdyt ?
> > > (asking because we had this conversation in the past)

maiqueb commented:
Something I missed saying before below:

> I discussed with Alona offline briefly, and in the past about it, I will start with updating the resync one, since the unit test should test that,
> once its ready we can discuss again.
> There are e2e tests that do check the full flow.

The fact that you had to revert code is proof that we cannot trust these end 2 end tests.

> The unit test checks the resync logic, if there resync is working as expected, the rest is for the e2e to check it according my view.
> 
> > > @AlonaKaplan wdyt ?
> > > (asking because we had this conversation in the past)

oshoval commented:
> The fact that you had to revert code is proof that we cannot trust these end 2 end tests.

There are false positives due to flaky CI, so sometimes real stuff evades.

I think we should focus on the PR scope, and can even create a follow PR
which have e2e bridge binding + guest agent, this scenario as well can be interesting,
for example because in the boot time, it might have not stable interface list (and not due to this PR).

There are more corner cases which I noticed but not related to this PR, so we will discuss them offline.
This PR and its predecessor address 3 BZ, there were lots of details.
Personally I believe in dividing stuff, unless mandatory.

oshoval commented:
Changes, added unit test for `func (s *AsyncAgentStore) GetInterfaceStatus()`
https://github.com/kubevirt/kubevirt/pull/5560/files#diff-510b183b2d5d079863708a79f37a82f552d485d8c3e6e10e5e2eefecfc946edaR134

At the current design of the PR, we can't mock the internals of `InterfacesStatus`
https://github.com/kubevirt/kubevirt/pull/5560/files#diff-98b8339212e612cd92ce33649e492336892a3ed649f368b181aad9aa2854138cR1373
but just domain manager, I don't think it deserve redesign.

oshoval commented:
/cc @EdDev

oshoval commented:
one of the failures of https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5560/pull-kubevirt-unit-test/1387742729437450240  (the GetJobStats one)
happens also on master, checking if the other as well

oshoval commented:
/retest

EdDev commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1367,6 +1369,17 @@ func (l *LibvirtDomainManager) GetGuestInfo() (v1.VirtualMachineInstanceGuestAge
+}
+
+// GetGuestOSInfo returns the Guest OS version and architecture
+func (l *LibvirtDomainManager) GetGuestOSInfo() api.GuestOSInfo {
```
Didn't we dropped the `get` previously?

oshoval replied:
On the other function `InterfacesStatus` not on this one, i can drop this one as well 

Referring to https://github.com/kubevirt/kubevirt/pull/5560#pullrequestreview-649883591
This is the only reason i didn't squash yet, i will squash at the end, as stated previously

EdDev replied:
> I didn't squash yet so it will be easier to see what are the new changes, i will squash at the end, as stated previously

I do not know why I need to bother myself with what was before, when it had a bug.
It makes me harder to review.
If you got a request to split from another reviewer, then fine. Otherwise, I do not understand how is this helpful.

oshoval replied:
It was helpful for me, i will squash as soon as possible

EdDev commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -83,6 +83,13 @@ var _ = Describe("Qemu agent poller", func() {
+			var agentStore = NewAsyncAgentStore()
+			agent := agentStore.GetInterfaceStatus()
+
+			Expect(agent == nil).To(Equal(true))
```
I do not thing `agent` fits here, how about `interfacesStatus`?

And the assert: `Expect(interfacesStatus).To(BeNil())`

oshoval commented:
/retest

EdDev commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1372,7 +1372,11 @@ func (l *LibvirtDomainManager) GetGuestInfo() (v1.VirtualMachineInstanceGuestAge
 func (l *LibvirtDomainManager) InterfacesStatus(domainInterfaces []api.Interface) []api.InterfaceStatus {
 	interfaces := l.agentData.GetInterfaceStatus()
-	return agentpoller.MergeAgentStatusesWithDomainData(domainInterfaces, interfaces)
+	if interfaces != nil {
+		return agentpoller.MergeAgentStatusesWithDomainData(domainInterfaces, interfaces)
+	}
```
Can this be embedded in `MergeAgentStatusesWithDomainData`?
As seen here, this is bug prone to assume the caller needs to treat this case.

oshoval replied:
Waiting with this to the sumup of what we are doing with the unit tests / e2e
more info here
https://github.com/kubevirt/kubevirt/pull/5560#issuecomment-830764134

oshoval replied:
The function does a conversion, in case its triggered, 
so it can't just return the input self, see please that there are 2 distinct input types
and one output type.

func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceStatuses []api.InterfaceStatus) []api.InterfaceStatus

EdDev commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -130,6 +130,28 @@ func (s *AsyncAgentStore) GetSysInfo() api.DomainSysInfo {
+// GetGuestOSInfo returns the Guest OS version and architecture
+func (s *AsyncAgentStore) GetGuestOSInfo() api.GuestOSInfo {
+	data, ok := s.store.Load(GET_OSINFO)
+	osinfo := api.GuestOSInfo{}
+	if ok {
+		osinfo = data.(api.GuestOSInfo)
+	}
+
+	return osinfo
```
This better be treated like the interfaces one.

EdDev reviewed (changes requested):
> In case there is no guest agent,
> The VMI interfaces status for bridge binding is
> determined by the pod ip (saved in virt-handler cache).

Even if the Guest Agent is there, the IP reported should be of the pod and not what the agent reports. So this sentence does not make sense to me.

I also believe there is little sense to have one commit that introduces a bug and then a second one that fixes it. Please squash them or split it in a different way.

EdDev reviewed (changes requested):
I will try to rephrase @maiqueb tests requests, at least as I understand it.

This change has introduces code with conditions that treat a specific scenario.
We are supposed to reflect the changes in unit tests that exercises the new code and cover all flow paths it introduced.
If I will try to be specific here, this means that the new methods added to `LibvirtDomainManager` and the conditions in `GetDomain` should be covered. Are they?
If the are not covered, please explain the complexity to do so, in order to determine what to do.
(I do not see any of the mentioned methods dependent on a timer)

oshoval commented:
> I will try to rephrase @maiqueb tests requests, at least as I understand it.
> 
> This change has introduces code with conditions that treat a specific scenario.
> We are supposed to reflect the changes in unit tests that exercises the new code and cover all flow paths it introduced.
> If I will try to be specific here, this means that the new methods added to `LibvirtDomainManager` and the conditions in `GetDomain` should be covered. Are they?
> If the are not covered, please explain the complexity to do so, in order to determine what to do.
> (I do not see any of the mentioned methods dependent on a timer)

It does depends on timer, the flow that is changed by this PR is triggered by a timer
https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/cache/cache.go#L340
It won't be triggered if not waited enough.

The resync unit test is handled here (this is the only test that waits for the resync timer) 
https://github.com/kubevirt/kubevirt/pull/5560/files#diff-6406f1444186b45ac44ceb8e369fd9b6ce0c9a4cc6f313bf80e815f0f33e705bR301
(non nil scenario)

From what I discussed not once with @AlonaKaplan 
Unit test should not test e2e flow

If we rewrite `LibvirtDomainManager InterfacesStatus` such that it will return nil, and the caller will determines if to call `MergeAgentStatusesWithDomainData` we will be able to mock on
on the other hand it slightly contradicts https://github.com/kubevirt/kubevirt/pull/5560#discussion_r624642556
and the simpleness of the current design, but i think that if we want to cover some flow in unit test this what we should do,

otherwise it means to write a unit test that check all the flow including the virt-handler vm.go logic

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1372,7 +1372,11 @@ func (l *LibvirtDomainManager) GetGuestInfo() (v1.VirtualMachineInstanceGuestAge
 	interfaces := l.agentData.GetInterfaceStatus()
-	return agentpoller.MergeAgentStatusesWithDomainData(domainInterfaces, interfaces)
+	if interfaces != nil {
+		return agentpoller.MergeAgentStatusesWithDomainData(domainInterfaces, interfaces)
```
I would add unit test here as well to make sure that in case `l.agentData.GetInterfaceStatus()` is need the `Merge...` is not invoked.

oshoval replied:
It means that we need to redesign, is it ok ?
because i can just mock InterfacesStatus but not its internals ?

AlonaKaplan replied:
No need to redesign, you just need to check the return value of `InterfacesStatus`. It should be nil.

AlonaKaplan commented on pkg/virt-launcher/virtwrap/cmd-server/server.go:
```diff
@@ -310,7 +310,10 @@ func (l *Launcher) GetDomain(_ context.Context, _ *cmdv1.EmptyRequest) (*cmdv1.D
 		domainObj.Status.OSInfo = l.domainManager.GetGuestOSInfo()
-		domainObj.Status.Interfaces = l.domainManager.InterfacesStatus(domainObj.Spec.Devices.Interfaces)
+		interfaces := l.domainManager.InterfacesStatus(domainObj.Spec.Devices.Interfaces)
+		if interfaces != nil {
```
I would add a unit test here to make sure that if `l.domainManager.InterfacesStatus(domainObj.Spec.Devices.Interfaces)` is `nil` the returned `domainObj.Status.Interfaces` is nil.

oshoval replied:
It should be left untouched, not nil
(in case it was nil it will of course stay nil)

it depends on the redesign, i will try to create a commit for it

AlonaKaplan replied:
We want to test the no guest agent scenario, so I would test it with `nil`. But of course the unit test for other values are welcome as well.

EdDev commented:
> > I will try to rephrase @maiqueb tests requests, at least as I understand it.
> > This change has introduces code with conditions that treat a specific scenario.
> > We are supposed to reflect the changes in unit tests that exercises the new code and cover all flow paths it introduced.
> > If I will try to be specific here, this means that the new methods added to `LibvirtDomainManager` and the conditions in `GetDomain` should be covered. Are they?
> > If the are not covered, please explain the complexity to do so, in order to determine what to do.
> > (I do not see any of the mentioned methods dependent on a timer)
> 
> It does depends on timer, the flow that is changed by this PR is triggered by a timer
> https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/cache/cache.go#L340
> It won't be triggered if not waited enough.

I did not ask to test the caller, I asked to test the method that was changed. That function has no timer.


> From what I discussed not once with @AlonaKaplan
> Unit test should not test e2e flow

As mentioned before, I have not asked to test the whole flow, only the methods that have been changed.

> If we rewrite `LibvirtDomainManager InterfacesStatus` such that it will return nil, and the caller will determines if to call `MergeAgentStatusesWithDomainData` we will be able to mock it.
> on the other hand it slightly contradicts [#5560 (comment)](https://github.com/kubevirt/kubevirt/pull/5560#discussion_r624642556)
> and the simpleness of the current design, but i think that if we want to cover the nil flow in unit test this what we should do,

I think you can balance between the comments such that you get the best result possible. Options always have pros and cons, the end result depends on the priority (coverage is a high priority) and technical ability to implement the needed.

oshoval commented:
Rebased

oshoval commented:
Addressed comments
squashed
unit test that were added are at
pkg/virt-launcher/virtwrap/cmd-server/server_test.go
pkg/virt-launcher/virtwrap/manager_test.go

oshoval commented:
Added the check if nil -> return nil to `MergeAgentStatusesWithDomainData`
according comment

oshoval commented:
Same test 5252 failed on those two failures
opened https://github.com/kubevirt/kubevirt/pull/5579 to quarantine it

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1623,17 @@ var _ = Describe("Manager", func() {
+			It("should return nil when no interfaces exists", func() {
+				agentStore := agentpoller.NewAsyncAgentStore()
+				manager, _ := NewLibvirtDomainManager(mockConn, testVirtShareDir, nil, 0, &agentStore, "/usr/share/OVMF")
+				libvirtmanager := manager.(*LibvirtDomainManager)
```
Why this is needed?

oshoval replied:
no need, done

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1623,17 @@ var _ = Describe("Manager", func() {
+
+		Context("on call to InterfacesStatus", func() {
+			It("should return nil when no interfaces exists", func() {
+				agentStore := agentpoller.NewAsyncAgentStore()
+				manager, _ := NewLibvirtDomainManager(mockConn, testVirtShareDir, nil, 0, &agentStore, "/usr/share/OVMF")
```
This seems to fit the setup fixture of the test, so it will be nice to see it in a `BeforeEach`.

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1623,17 @@ var _ = Describe("Manager", func() {
+				interfaces := libvirtmanager.InterfacesStatus(nil)
+				Expect(interfaces).To(BeNil())
+			})
+		})
```
I'm missing here coverage for `GetGuestOSInfo` and `InterfacesStatus` with a non empty domain interface list as an argument.

oshoval replied:
GetGuestOSInfo with info is covered in the resync test, unless it needs a test here as well 
see please comment
https://github.com/kubevirt/kubevirt/pull/5560#discussion_r626284924

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -214,6 +214,10 @@ func parseAgent(agentReply string) (AgentInfo, error) {
 
 // MergeAgentStatusesWithDomainData merge QEMU interfaces with agent interfaces
 func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceStatuses []api.InterfaceStatus) []api.InterfaceStatus {
+	if interfaceStatuses == nil {
```
So i prefer to do it in a separate PR if possible, and focus this PR on the required changes.

oshoval replied:
I mean that i will keep the callers check it, this was the state before this PR
i just added a caller, i will let the new caller check it, and refactor can be done on a separate PR
will just revert this push
https://github.com/kubevirt/kubevirt/compare/aebad1086fbf6453b44f3b74ed6a3a2cae38aa93..4c956b5f0641d5ca9cd362b2bc0004557fb919a1

EdDev replied:
As long as you will cover the branching of your new caller, do what you prefer.

oshoval commented on pkg/virt-handler/cache/cache_test.go:
```diff
@@ -314,6 +324,8 @@ var _ = Describe("Domain informer", func() {
 
 			Expect(ok).To(BeTrue())
 			Expect(val).To(Equal("some-value"))
+			Expect(eventDomain.Status.OSInfo.Name).To(Equal("fedora"))
+			Expect(eventDomain.Status.Interfaces[0].Name).To(Equal("ethx"))
```
This test is the one that unit tests the resync, and waits for the timer of it 
i extended the resync, hence extended the test
you can see the PR that introduced the resync + the unit test of it at
https://github.com/kubevirt/kubevirt/pull/3893

This btw also why i didn't test some of the flows of non empty, because we already have them here
referring for example to your comment
https://github.com/kubevirt/kubevirt/pull/5560#discussion_r626266763
unless its need to be doubled

Btw this change was already in the original PR that was reverted.

oshoval replied:
We check that the data is added to the resync flow is working.
I think it worth to keep it, we can add other unit tests as well if needed, but i wouldn't drop it
wdyt?

EdDev replied:
The resync flow is working because it is tested with some arbitrary/sample data. If we need to test all data options, then we would not be able to maintain such a huge test.

Maybe I can illustrate it like this:
A is the resync mechanism.
B is the data population.

Production flow is: A ==>> B
You changed B.
A was not changed, therefore it requires no special coverage.
B was changes, it requires coverage of the new scenarios.

Testing A to validate B changes is not advised, it overloads A with indirect implications which come from B.

oshoval replied:
ok will drop

oshoval commented on pkg/virt-launcher/virtwrap/cmd-server/server_test.go:
```diff
@@ -123,12 +123,16 @@ var _ = Describe("Virt remote commands", func() {
 			Expect(exists).To(BeTrue())
 			Expect(domain).ToNot(Equal(nil))
 			Expect(domain.ObjectMeta.Name).To(Equal("testvmi1"))
+			Expect(domain.Status.OSInfo).To(Equal(api.GuestOSInfo{}))
+			Expect(domain.Status.Interfaces).To(BeNil())
```
see please 
https://github.com/kubevirt/kubevirt/pull/5560#discussion_r626284924

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -130,6 +130,27 @@ func (s *AsyncAgentStore) GetSysInfo() api.DomainSysInfo {
+		return data.([]api.InterfaceStatus)
+	}
+
+	return nil
```
see please 
https://github.com/kubevirt/kubevirt/pull/5560#discussion_r626284924

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -130,6 +130,27 @@ func (s *AsyncAgentStore) GetSysInfo() api.DomainSysInfo {
+}
+
+// GetGuestOSInfo returns the Guest OS version and architecture
+func (s *AsyncAgentStore) GetGuestOSInfo() *api.GuestOSInfo {
```
It does, in other unit tests, one test it with data and one without
see please other comments, and i will recheck unless it needs its own minimized unit test as well

oshoval commented on pkg/virt-launcher/virtwrap/cmd-server/server.go:
```diff
@@ -308,7 +308,16 @@ func (l *Launcher) GetDomain(_ context.Context, _ *cmdv1.EmptyRequest) (*cmdv1.D
+		if osInfo != nil {
+			domainObj.Status.OSInfo = *osInfo
+		}
+		interfaces := l.domainManager.InterfacesStatus(domainObj.Spec.Devices.Interfaces)
+		if interfaces != nil {
```
can u explain please what u mean ?
i need the result of it twice, without calling it twice
one in the comparison and one on the assignment if its not nil

(same for the other comment)

EdDev replied:
`if f := foo(); f != nil { a.f = f }`

oshoval commented:
first push - address comments
2nd push - rebase

oshoval commented on pkg/virt-launcher/virtwrap/BUILD.bazel:
```diff
@@ -61,6 +61,7 @@ go_test(
         "//pkg/handler-launcher-com/cmd/v1:go_default_library",
         "//pkg/util/net/ip:go_default_library",
         "//pkg/virt-handler/cmd-client:go_default_library",
+        "//pkg/virt-launcher/virtwrap/agent-poller:go_default_library",
```
maybe need to drop this, checking

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+
+				agentStore.Store(agentpoller.GET_OSINFO, fakeInfo)
+				osInfo := libvirtmanager.GetGuestOSInfo()
+				Expect(reflect.DeepEqual(&fakeInfo, osInfo)).To(BeTrue())
```
Why do you need to use `reflect` here? Isn't a simple struct equality enough?

oshoval replied:
Its a pointer, i can deref instead, but DeepEqual does it
wdyt?
lets first answer if we keep it as pointer (see please https://github.com/kubevirt/kubevirt/pull/5560#discussion_r629074725)

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+					Name: "TestGuestOSName",
+				}
+
+				agentStore.Store(agentpoller.GET_OSINFO, fakeInfo)
```
This is part of the test preparation, so please place it in the same block as the `fakeInfo` (i.e. no newline between them, newline after).

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+		Context("on call to InterfacesStatus", func() {
+			var libvirtmanager DomainManager
+			var agentStore agentpoller.AsyncAgentStore
+			domInterfaces := []api.Interface{
```
Please be consistent here, if you call one info "fake", do the same to all. I do not really understand why it needs to be called a fake, as it is a regular data, but I do not really mind much.

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+			})
+
+			It("should return nil when no interfaces exists in the cache, nor as argument", func() {
+				interfaces := libvirtmanager.InterfacesStatus(nil)
+				Expect(interfaces).To(BeNil())
```
nit: can be inline.

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+			})
+
+			It("should return nil when no interfaces exists in the cache", func() {
+				interfaces := libvirtmanager.InterfacesStatus(domInterfaces)
+				Expect(interfaces).To(BeNil())
```
can be inline

oshoval replied:
less reable imo,
personally its much easier to read in two lines, what we get and what we do with it
but ok, already did.. (on next push)

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+
+			It("should return merged list when interfaces exists on both the cache and argument", func() {
+				expectedResult := []api.InterfaceStatus{
+					{
+						Name: "eth2",
+						Mac:  "00:00:00:00:00:02",
+					},
```
Please express that this is in fact `fakeInterfaces[0]` by ref it here.

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+						Name: "eth2",
+						Mac:  "00:00:00:00:00:02",
+					},
+					{
+						Name: "eth1",
+						Mac:  "00:00:00:00:00:01",
+					},
```
Same, show it is from `domInterfaces[0]` (with the needed translation).

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1622,6 +1624,95 @@ var _ = Describe("Manager", func() {
+
+			It("should return merged list when interfaces exists on the cache only", func() {
+				expectedResult := []api.InterfaceStatus{
+					{
+						Name: "eth2",
+						Mac:  "00:00:00:00:00:02",
+					},
```
same, express by ref from where this comes from.

oshoval commented:
Rebase

oshoval commented on pkg/virt-launcher/virtwrap/cmd-server/server_test.go:
```diff
@@ -118,17 +118,45 @@ var _ = Describe("Virt remote commands", func() {
 		})
 
-		It("should list domains", func() {
+		It("should list domains when no cache info exists", func() {
```
yep, just wrote cache because it looks in the cache that holds it, will change to guest agent info

oshoval commented:
Addressed (the resolved) comments

EdDev commented on pkg/virt-launcher/virtwrap/cmd-server/server_test.go:
```diff
@@ -118,17 +118,45 @@ var _ = Describe("Virt remote commands", func() {
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(exists).To(BeTrue())
+			Expect(domain).ToNot(Equal(nil))
```
I do not know where from you took this assumption.
`BeNil` is checking nil, not an empty list or an empty anything.

EdDev replied:
What you showed is wrong.
One `nil` is of a specific type and the other has no type. I do not even know how you managed to reach such a result.

oshoval replied:
in case you have a function that returns type []api.InterfaceStatus` and actually return nil, this what will happen

EdDev replied:
Per what I know, it will return a list type with a nil value, not a non-type nil like you claim.
It is true that you can compare any type nil value to a non-type nil, but that is surely an exception here to allow some level of generics.
I would conclude with a snippet:
```golang
It("foo", func() {
        var list []string
        Expect(list).To(BeNil())

        var list2 = []string{}
        Expect(list2).NotTo(BeNil())
})
```

oshoval replied:
Try please `To(Equal(nil))` and see the assert that i shown

EdDev commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -83,6 +84,54 @@ var _ = Describe("Qemu agent poller", func() {
+			agentStore.Store(GET_OSINFO, fakeInfo)
+			osInfo := agentStore.GetGuestOSInfo()
+
+			Expect(reflect.DeepEqual(&fakeInfo, osInfo)).To(BeTrue())
```
I have not commented on it being a pointer or not.
I asked why you cannot use a simple struct equality.
(if something is a pointer, you can use `*` to evaluate its content)

EdDev reviewed (approved):
Thank you!

oshoval commented on pkg/virt-launcher/virtwrap/cmd-server/server_test.go:
```diff
@@ -118,17 +118,45 @@ var _ = Describe("Virt remote commands", func() {
+
+			Expect(exists).To(BeTrue())
+			Expect(domain).ToNot(BeNil())
+			Expect(domain.ObjectMeta.Name).To(Equal("testvmi1"))
```
Yes and no, the specific test checks the osInfo and the interfaces,
other tests check the VM name, so its a copy paste, and the same state on the other tests

I can either leave it as is, update it, or remove it, up to you
thanks

EdDev replied:
Are you reasoning that the other tests are better?
This is not a blocker, but lets try to use valid reasoning.

AlonaKaplan commented on pkg/virt-handler/cache/cache_test.go:
```diff
@@ -284,12 +290,16 @@ var _ = Describe("Domain informer", func() {
 			newDomain.ObjectMeta.Labels["some-label"] = "some-value"
 			domainManager.EXPECT().ListAllDomains().Return([]*api.Domain{newDomain}, nil)
+			domainManager.EXPECT().GetGuestOSInfo().Return(&api.GuestOSInfo{Name: "fedora"})
+			domainManager.EXPECT().InterfacesStatus(newDomain.Spec.Devices.Interfaces).Return([]api.InterfaceStatus{api.InterfaceStatus{Name: "ethx"}})
```
Isn't [this](https://github.com/kubevirt/kubevirt/pull/5428/commits/3e9ba98a98acc41134788bb1e85bfca9eeeea9e0#diff-6406f1444186b45ac44ceb8e369fd9b6ce0c9a4cc6f313bf80e815f0f33e705bR327) part missing?

oshoval replied:
Edy asked to remove it
See here please why
https://github.com/kubevirt/kubevirt/pull/5560#discussion_r626322993

AlonaKaplan replied:
I see, so please drop the "special" mocking here ("fedora", "ethx"). It is confusing.

oshoval commented:
Addressed (both) comments

AlonaKaplan commented:
/approve

oshoval commented:
Thanks

/retest

stu-gott commented:
/cherry-pick release-0.36
/cherry-pick release-0.41

stu-gott commented:
/cherry-pick release-0.41

acardace commented:
Manual cherry-pick created in https://github.com/kubevirt/kubevirt/pull/5655
